### PR TITLE
NMS-13073: BsfDetector documenation

### PR DESCRIPTION
**Pull request by mfuhrmann**
### All Contributors

* [x] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [x] Have you made a comment in that issue which points back to this PR?
* [x] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [x] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-13073

### CODE REVIEW

**Comment by mfuhrmann** [comment]
I've fixed all the requests mentioned in https://github.com/OpenNMS/opennms/pull/3349.
Also, I've tested all the example scripts from the BsfMonitor docs. They work absolutely identical in the detector :+1:

BUT I had to remove this part from the scripts:
`bsf_monitor.log("ERROR", "Starting MinimalGroovy", null);`
It seems only to work for the monitor. I've tried `bsf_detector.log...` without success.

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
This is great, Marcel!

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  return "OK";
+} else {
+  results.put("reason", "file does not exist");
+  return "NOK";
+}
+----
+
+=== Example Groovy
+
+._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `eval`
```
[the remark this answers is not in the document]
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc]: Succeeded!

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+<detector name="MinimalGroovy" class="org.opennms.netmgt.provision.detector.bsf.BSFDetector">
+    <parameter key="bsfEngine" value="org.codehaus.groovy.bsf.GroovyEngine"/>
+    <parameter key="langClass" value="groovy"/>
+    <parameter key="fileName" value="/opt/opennms/etc/scripts/MinimalGroovy.groovy"/>
+    <parameter key="runType" value="exec"/>
+</detector>
+
+----
+
+.BeanShell example `MinimalGroovy.groovy` script file for `run-type` `exec`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:132]:
```suggestion
._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `exec`
```

[the remark this answers is not in the document]
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:132]: yes

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+----
+<detector name="MinimalGroovy" class="org.opennms.netmgt.provision.detector.bsf.BSFDetector">
+    <parameter key="bsfEngine" value="org.codehaus.groovy.bsf.GroovyEngine"/>
+    <parameter key="langClass" value="groovy"/>
+    <parameter key="fileName" value="/opt/opennms/etc/scripts/MinimalGroovy.groovy"/>
+    <parameter key="runType" value="eval"/>
+</detector>
+----
+
+.BeanShell example `MinimalGroovy.groovy` script file for `run-type` `eval`
```
[the remark this answers is not in the document]
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:108]:
```suggestion
._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `eval`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 1 above)
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc]:
```suggestion
._Groovy_ example for `run-type` `eval`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+File testFile = new File("/tmp/TestFile");
+if (testFile.exists()) {
+  return "OK";
+} else {
+  results.put("reason", "file does not exist");
+  return "NOK";
+}
+----
+
+._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `exec`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc]:
```suggestion
._Groovy_ example for `run-type` `exec`
```

**Comment by mfuhrmann** [comment]
@Bonrob2 Thanks for checking. Not sure why 2 of my suggestions are outdated now.

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 4 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc]: @mfuhrmann . Strange. Just need to remove the italics around "Groovy".

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  return "OK";
+} else {
+  results.put("reason", "file does not exist");
+  return "NOK";
+}
+----
+
+=== Groovy Example 
+
+.Groovy example `MinimalGroovy.groovy` script file for `run-type` `eval`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:97]:
```suggestion
.Groovy example for `run-type` `eval`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+----
+<detector name="MinimalGroovy" class="org.opennms.netmgt.provision.detector.bsf.BSFDetector">
+    <parameter key="bsfEngine" value="org.codehaus.groovy.bsf.GroovyEngine"/>
+    <parameter key="langClass" value="groovy"/>
+    <parameter key="fileName" value="/opt/opennms/etc/scripts/MinimalGroovy.groovy"/>
+    <parameter key="runType" value="eval"/>
+</detector>
+----
+
+._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `eval`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:108]:
```suggestion
.Groovy example `MinimalGroovy.groovy` script file for `run-type` `eval`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+File testFile = new File("/tmp/TestFile");
+if (testFile.exists()) {
+  return "OK";
+} else {
+  results.put("reason", "file does not exist");
+  return "NOK";
+}
+----
+
+.Groovy example `MinimalGroovy.groovy` script file for `run-type` `exec`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:120]:
```suggestion
.Groovy example for `run-type` `exec`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+<detector name="MinimalGroovy" class="org.opennms.netmgt.provision.detector.bsf.BSFDetector">
+    <parameter key="bsfEngine" value="org.codehaus.groovy.bsf.GroovyEngine"/>
+    <parameter key="langClass" value="groovy"/>
+    <parameter key="fileName" value="/opt/opennms/etc/scripts/MinimalGroovy.groovy"/>
+    <parameter key="runType" value="exec"/>
+</detector>
+
+----
+
+._Groovy_ example `MinimalGroovy.groovy` script file for `run-type` `exec`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:132]:
```suggestion
.Groovy example `MinimalGroovy.groovy` script file for `run-type` `exec`
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+}
+----
+
+=== Example Jython
+
+To use the Jython (Java implementation of Python) language an additional library is required.
+Copy a compatible `jython-x.y.z.jar` into the `opennms/lib` folder and restart {page-component-title}.
+That makes Jython available for the BSF Detector.
+
+.Jython example `MinimalJython.py` script file for `run-type` `exec`
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:150]:
```suggestion
.Jython example for `run-type` `exec`
```

**Review by Bonrob2 — APPROVED** [review]

**Review by Bonrob2 — APPROVED** [review]

**Review by Bonrob2 — APPROVED** [review]

**Review by Bonrob2 — APPROVED** [review]

**Review by Bonrob2 — APPROVED** [review]

**Comment by Bonrob2** [comment]
LGTM!

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 10]
```diff
@@ -0,0 +1,175 @@
+= BSF Detector
+
+The BSF Detector lets you run arbitrary scripts using the Bean Scripting Framework to determine the availability of a service.
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:3]:
```suggestion
The BSF Detector lets you run arbitrary scripts using the Bean Scripting Framework to determine the existence of a service.
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+This can be a script written in any of the languages BSF supports.
+The only requirement is that the script returns the string `OK` if it passes.
+
+Unlike the BSF Monitor, the BSF Detector only passes in the `map`, `ip_addr`, `svc_name`, and `results` variables as beans to the script.
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:7]:
```suggestion
The BSF Detector passes the `map`, `ip_addr`, `svc_name`, and `results` variables as beans to the script.
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+== Configuration and Use
+
+.Parameters for the BSF detector
+[options="header, autowidth"]
+
+|===
+| Parameter        | Description                                                                                            | Required | Default value
+| `fileName`       | The path to the file on the OpenNMS server which contains the BSF-compatible script to run.            | optional |
+| `langClass`      | The short name of the language the script is written in. Choices: Jython, BeanShell, Groovy, etc.     | optional |
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:24]:
```suggestion
| `langClass`      | The short name of the language the script is written in. Choices: `Jython`, `BeanShell`, `Groovy`, etc.     | optional | 
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+|===
+| Parameter        | Description                                                                                            | Required | Default value
+| `fileName`       | The path to the file on the OpenNMS server which contains the BSF-compatible script to run.            | optional |
+| `langClass`      | The short name of the language the script is written in. Choices: Jython, BeanShell, Groovy, etc.     | optional | 
+| `bsfEngine`      | The BSF engine to use when running the script. Choices:
+
+* bsh.util.BeanShellBSFEngine
+* org.codehaus.groovy.bsf.GroovyEngine
+* org.apache.bsf.engines.jython.JythonEngine                                                                                | optional |
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:29]: Lists in table does not work. The choices you have to use in your script, I would expect it also in `code` formatting.

Screenshot 2021-04-19 at 09 46 13

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+| `bsfEngine`      | The BSF engine to use when running the script. Choices:
+
+* bsh.util.BeanShellBSFEngine
+* org.codehaus.groovy.bsf.GroovyEngine
+* org.apache.bsf.engines.jython.JythonEngine                                                                                | optional | 
+| `fileExtensions` | A comma-separated list of file extensions to consider a valid script for this BSF instance.            | optional | 
+| `runType`        | Which way to run the BSF script. Choices:
+
+* `eval` - expects the script to print the OK/NOK status for validation
+* `exec` - expects the script to manipulate the provided results hashmap directly                                           | optional | `eval`
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:34]: List does not work, same here

Screenshot 2021-04-19 at 09 47 06

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 12 above)
Comment (indigo423) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:24]: It seems like the options you can use are lower case, right?

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+The BSF detector supports many languages; the following table provides the required setup for commonly used languages.
+
+.BSF language setups
+[options="header, %autowidth"]
+|===
+| Language                            | lang-class  | bsf-engine                                    | required library
+| http://www.beanshell.org[BeanShell] | beanshell | `bsh.util.BeanShellBSFEngine`                 | supported by default
+| http://groovy.codehaus.org[Groovy]  | groovy  | `org.codehaus.groovy.bsf.GroovyEngine`        | +groovy-all-[version].jar+
+| http://www.jython.org[Jython]       |  jython   | `org.apache.bsf.engines.jython.JythonEngine`  | +jython-[version].jar+
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:64]:
```suggestion
| http://www.jython.org[Jython]       | jython   | `org.apache.bsf.engines.jython.JythonEngine`  | +jython-[version].jar+
```

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+if (File("/tmp/TestFile").exists()):
+        results.put("status", "OK")
+else:
+        results.put("reason", "file does not exist")
+        results.put("status", "NOK")
+----
+
+NOTE: We have to use `run-type` `exec` here because Jython chokes on the +import+ keyword in `eval` mode.
+
+NOTE: As proof that this is really _Python_, notice the substitution of _Python's_ +None+ value for Java's +null+ in the log call.
```
Comment (indigo423) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:175]: Do we really need the proof NOTE?

**Review by indigo423 — COMMENTED** [review]
Only minor things

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 12 above)
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:24]: Never saw upper case options TBH.

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 13 above)
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc]: Now it works: https://github.com/OpenNMS/opennms/blob/82b7042d726b29f3f15001ae7c88d9283d6869b2/docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 14 above)
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:34]: Now it works: https://github.com/OpenNMS/opennms/blob/82b7042d726b29f3f15001ae7c88d9283d6869b2/docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc (hunk 16 above)
Comment (mfuhrmann) [reply docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:175]: I don't know TBH. For me it looks like it does not help much.

File: docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/provisioning/detectors/BsfDetector.adoc:175]:
```suggestion
```

**Comment by RangerRick** [comment]
FYI, now that the doc changes have been merged to `foundation-2021`, I retargeted this, looks like it needs some conflict resolution though

**Comment by mfuhrmann** [comment]
I will create a new PR for this since I don't know how to fix those conflicts :joy:

**Pull request closed without merging**